File: Common/InMotionGIT.Common - Copy/Helpers/AssemblyHandler.vb
Imports System.Reflection
Imports System.Text
Imports System.Web.Hosting

Namespace Helpers

    Public Class AssemblyHandler

        ''' <summary>
        ''' Initializes a new instance of the AssemblyInfoCalling class.
        ''' </summary>
        ''' <param name="traceLevel">The trace level needed to get correct assembly
        ''' - will need to adjust based on where you put these classes in your project(s).</param>
        Public Sub New(Optional traceLevel As Integer = 4)
            '----------------------------------------------------------------------
            ' Default to "3" as the number of levels back in the stack trace to get the
            '  correct assembly for "calling" assembly
            '----------------------------------------------------------------------
            StackTraceLevel = traceLevel
        End Sub

        '----------------------------------------------------------------------
        ' Standard assembly attributes
        '----------------------------------------------------------------------
        Public ReadOnly Property Company() As String
            Get
                Return GetCallingAssemblyAttribute(Of AssemblyCompanyAttribute)(Function(a) a.Company)
            End Get
        End Property

        Public ReadOnly Property Product() As String
            Get
                Return GetCallingAssemblyAttribute(Of AssemblyProductAttribute)(Function(a) a.Product)
            End Get
        End Property

        Public ReadOnly Property Copyright() As String
            Get
                Return GetCallingAssemblyAttribute(Of AssemblyCopyrightAttribute)(Function(a) a.Copyright)
            End Get
        End Property

        Public ReadOnly Property Trademark() As String
            Get
                Return GetCallingAssemblyAttribute(Of AssemblyTrademarkAttribute)(Function(a) a.Trademark)
            End Get
        End Property

        Public ReadOnly Property Title() As String
            Get
                Return GetCallingAssemblyAttribute(Of AssemblyTitleAttribute)(Function(a) a.Title)
            End Get
        End Property

        Public ReadOnly Property Description() As String
            Get
                Return GetCallingAssemblyAttribute(Of AssemblyDescriptionAttribute)(Function(a) a.Description)
            End Get
        End Property

        Public ReadOnly Property Configuration() As String
            Get
                Return GetCallingAssemblyAttribute(Of AssemblyDescriptionAttribute)(Function(a) a.Description)
            End Get
        End Property

        Public ReadOnly Property FileVersion() As String
            Get
                Return GetCallingAssemblyAttribute(Of AssemblyFileVersionAttribute)(Function(a) a.Version)
            End Get
        End Property

        '----------------------------------------------------------------------
        ' Version attributes
        '----------------------------------------------------------------------
        Public Shared ReadOnly Property Version() As Version
            Get
                '----------------------------------------------------------------------
                ' Get the assembly, return empty if null
                '----------------------------------------------------------------------
                Dim assembly As Assembly = GetAssembly(StackTraceLevel)
                Return If(assembly Is Nothing, New Version(), assembly.GetName().Version)
            End Get
        End Property

        Public ReadOnly Property VersionFull() As String
            Get
                Return Version.ToString()
            End Get
        End Property

        Public ReadOnly Property VersionMajor() As String
            Get
                Return Version.Major.ToString()
            End Get
        End Property

        Public ReadOnly Property VersionMinor() As String
            Get
                Return Version.Minor.ToString()
            End Get
        End Property

        Public ReadOnly Property VersionBuild() As String
            Get
                Return Version.Build.ToString()
            End Get
        End Property

        Public ReadOnly Property VersionRevision() As String
            Get
                Return Version.Revision.ToString()
            End Get
        End Property

        '----------------------------------------------------------------------
        ' Set how deep in the stack trace we're looking - allows for customized changes
        '----------------------------------------------------------------------
        Public Shared Property StackTraceLevel() As Integer
            Get
                Return m_StackTraceLevel
            End Get
            Set
                m_StackTraceLevel = Value
            End Set
        End Property

        Private Shared m_StackTraceLevel As Integer

        ''' <summary>
        ''' Gets the calling assembly attribute.
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="value">The value.</param>
        ''' <example>return GetCallingAssemblyAttribute&lt;AssemblyCompanyAttribute&gt;(a => a.Company);</example>
        ''' <returns></returns>
        Private Function GetCallingAssemblyAttribute(Of T As Attribute)(value As Func(Of T, String)) As String
            '----------------------------------------------------------------------
            ' Get the assembly, return empty if null
            '----------------------------------------------------------------------
            Dim assembly As Assembly = GetAssembly(StackTraceLevel)
            If assembly Is Nothing Then
                Return String.Empty
            End If

            '----------------------------------------------------------------------
            ' Get the attribute value
            '----------------------------------------------------------------------
            Dim attribute__1 As T = DirectCast(Attribute.GetCustomAttribute(assembly, GetType(T)), T)
            Return value.Invoke(attribute__1)
        End Function

        ''' <summary>
        ''' Go through the stack and gets the assembly
        ''' </summary>
        ''' <param name="stackTraceLevel">The stack trace level.</param>
        ''' <returns></returns>
        Private Shared Function GetAssembly(stackTraceLevel As Integer) As Assembly
            '----------------------------------------------------------------------
            ' Get the stack frame, returning null if none
            '----------------------------------------------------------------------
            Dim stackTrace As New StackTrace()
            Dim stackFrames As StackFrame() = stackTrace.GetFrames()
            If stackFrames Is Nothing Then
                Return Nothing
            End If

            '----------------------------------------------------------------------
            ' Get the declaring type from the associated stack frame, returning null if nonw
            '----------------------------------------------------------------------
            Dim declaringType = stackFrames(stackTraceLevel).GetMethod().DeclaringType
            If declaringType Is Nothing Then
                Return Nothing
            End If

            '----------------------------------------------------------------------
            ' Return the assembly
            '----------------------------------------------------------------------
            Dim assembly = declaringType.Assembly
            Return assembly
        End Function

        Friend Shared Function MethodName(currentMethod As String) As String
            Dim result As New StringBuilder
            Dim stackTrace As New StackTrace()
            Dim stackFrames As StackFrame() = stackTrace.GetFrames()
            Dim Index = stackFrames.[Select](Function(elem, indexInternal) New With {elem, indexInternal}).First(Function(p) (CType(p.elem, StackFrame)).GetMethod().Name.ToLower().Equals(currentMethod)).indexInternal + 1
            If stackFrames.IsNotEmpty AndAlso stackFrames.Length >= Index Then
                Dim stackCall As StackFrame = stackFrames(Index)
                With stackCall
                    If .GetMethod().IsNotEmpty AndAlso .GetMethod().Name.IsNotEmpty Then
                        result.AppendLine(String.Format("<<I>>Method:'{0}'", .GetMethod().Name))
                        Dim [nameSpace] As String = String.Empty
                        If .GetMethod().ReflectedType.IsNotEmpty AndAlso .GetMethod().ReflectedType.FullName.IsNotEmpty Then
                            Dim className As String = .GetMethod().ReflectedType.FullName
                            [nameSpace] = String.Format("{0}.{1}", className, .GetMethod().Name)
                        End If
                        If [nameSpace].IsEmpty Then
                            [nameSpace] = "Empty"
                        End If
                        result.AppendLine(String.Format("<<I>>Namespace:'{0}'", [nameSpace]))
                    Else
                        result.AppendLine(String.Format("<<I>>Method:'{0}'", "Empty"))
                    End If
                    result.AppendLine(String.Format("<<I>>Row/Column:'{0}'/'{1}'", .GetFileLineNumber, .GetFileColumnNumber))
                End With
            End If
            Return result.ToString
        End Function

        Public Shared Function GetFrameProcess(Index As Integer) As String
            Dim result As New StringBuilder
            Dim stackTrace As New StackTrace()
            Dim stackFrames As StackFrame() = stackTrace.GetFrames()
            If stackFrames.IsNotEmpty AndAlso stackFrames.Length >= Index Then
                Dim stackCall As StackFrame = stackFrames(Index)
                With stackCall
                    If .GetMethod().IsNotEmpty AndAlso .GetMethod().Name.IsNotEmpty Then
                        result.AppendLine(String.Format("<<I>>Method:'{0}'", .GetMethod().Name))
                        Dim [nameSpace] As String = String.Empty
                        If .GetMethod().ReflectedType.IsNotEmpty AndAlso .GetMethod().ReflectedType.FullName.IsNotEmpty Then
                            Dim className As String = .GetMethod().ReflectedType.FullName
                            [nameSpace] = String.Format("{0}.{1}", className, .GetMethod().Name)
                        End If
                        If [nameSpace].IsEmpty Then
                            [nameSpace] = "Empty"
                        End If
                        result.AppendLine(String.Format("<<I>>Namespace:'{0}'", [nameSpace]))
                    Else
                        result.AppendLine(String.Format("<<I>>Method:'{0}'", "Empty"))
                    End If
                    result.AppendLine(String.Format("<<I>>Row/Column:'{0}'/'{1}'", .GetFileLineNumber, .GetFileColumnNumber))
                End With
            End If
            Return result.ToString
        End Function

        Public Shared Function GetFrameProcessFullName(Index As Integer) As String
            Dim result As String = String.Empty
            Dim stackTrace As New StackTrace()
            Dim stackFrames As StackFrame() = stackTrace.GetFrames()
            If stackFrames.IsNotEmpty AndAlso stackFrames.Length >= Index Then
                Dim stackCall As StackFrame = stackFrames(Index)
                With stackCall
                    If .GetMethod().IsNotEmpty AndAlso .GetMethod().Name.IsNotEmpty Then
                        result = .GetMethod().Name
                        Dim [nameSpace] As String = String.Empty
                        If .GetMethod().ReflectedType.IsNotEmpty AndAlso .GetMethod().ReflectedType.FullName.IsNotEmpty Then
                            Dim className As String = .GetMethod().ReflectedType.FullName
                            [nameSpace] = String.Format("{0}.{1}", className, .GetMethod().Name)
                        End If
                        If [nameSpace].IsEmpty Then
                            [nameSpace] = "Empty"
                        End If
                        result = [nameSpace]
                    End If
                End With
            End If
            Return result
        End Function

        ''' <summary>
        ''' Obtiene el nombre del usuario de windows.
        ''' </summary>
        ''' <returns></returns>
        Public Shared Function GetUserNameWindows() As String
            Dim result As String = String.Empty
            If Not HostingEnvironment.IsHosted Then
                result = Environment.UserName
            End If
            Return result
        End Function

        ''' <summary>
        ''' Obtiene el nombre de la maquina.
        ''' </summary>
        ''' <returns></returns>
        Public Shared Function GetUserNameMachine() As String
            Dim result As String = String.Empty
            If Not HostingEnvironment.IsHosted Then
                result = Environment.MachineName
            End If
            Return result
        End Function

        ''' <summary>
        ''' Obtiene el nombre de la sistema operativo.
        ''' </summary>
        ''' <returns></returns>
        Public Shared Function GetOS() As String
            Dim result As String = String.Empty
            Try
                If Not HostingEnvironment.IsHosted Then
                    result = New Microsoft.VisualBasic.Devices.ComputerInfo().OSFullName
                End If
            Catch ex As Exception
                LogHandler.ErrorLog("InMotionGIT.Common.Helpers.AssemblyHandler", "GetOS", ex)
                result = "Unknown"
            End Try

            Return result
        End Function

        ''' <summary>
        ''' Obtiene el nombre de la sistema operativo.
        ''' </summary>
        ''' <returns></returns>
        Public Shared Function GetNameApplication() As String
            Dim result As String = String.Empty
            If Not HostingEnvironment.IsHosted Then
                Try
                    result = (New AssemblyHandler(5)).Title
                    If result.IsEmpty Then
                        result = Assembly.GetCallingAssembly.GetName().Name
                    End If
                Catch ex As Exception
                    result = "Name not found"
                End Try
            End If
            Return result
        End Function

        Public Shared Function GetVersionApplication() As String
            Dim result As String = String.Empty
            If Not HostingEnvironment.IsHosted Then
                Try
                    result = (New AssemblyHandler(5)).VersionFull
                    If result.IsEmpty Then
                        result = Assembly.GetCallingAssembly.GetName().Version.ToString
                    End If
                Catch ex As Exception
                    result = "-1"
                End Try
            End If
            Return result
        End Function

    End Class

End Namespace

File: Common/InMotionGIT.Common/Helpers/AssemblyHandler.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Web.Hosting;
using InMotionGIT.Common.Extensions;

namespace InMotionGIT.Common.Helpers
{

    public class AssemblyHandler
    {

        /// <summary>
        /// Initializes a new instance of the AssemblyInfoCalling class.
        /// </summary>
        /// <param name="traceLevel">The trace level needed to get correct assembly
        /// - will need to adjust based on where you put these classes in your project(s).</param>
        public AssemblyHandler(int traceLevel = 4)
        {
            // ----------------------------------------------------------------------
            // Default to "3" as the number of levels back in the stack trace to get the
            // correct assembly for "calling" assembly
            // ----------------------------------------------------------------------
            StackTraceLevel = traceLevel;
        }

        // ----------------------------------------------------------------------
        // Standard assembly attributes
        // ----------------------------------------------------------------------
        public string Company
        {
            get
            {
                return GetCallingAssemblyAttribute<AssemblyCompanyAttribute>(a => a.Company);
            }
        }

        public string Product
        {
            get
            {
                return GetCallingAssemblyAttribute<AssemblyProductAttribute>(a => a.Product);
            }
        }

        public string Copyright
        {
            get
            {
                return GetCallingAssemblyAttribute<AssemblyCopyrightAttribute>(a => a.Copyright);
            }
        }

        public string Trademark
        {
            get
            {
                return GetCallingAssemblyAttribute<AssemblyTrademarkAttribute>(a => a.Trademark);
            }
        }

        public string Title
        {
            get
            {
                return GetCallingAssemblyAttribute<AssemblyTitleAttribute>(a => a.Title);
            }
        }

        public string Description
        {
            get
            {
                return GetCallingAssemblyAttribute<AssemblyDescriptionAttribute>(a => a.Description);
            }
        }

        public string Configuration
        {
            get
            {
                return GetCallingAssemblyAttribute<AssemblyDescriptionAttribute>(a => a.Description);
            }
        }

        public string FileVersion
        {
            get
            {
                return GetCallingAssemblyAttribute<AssemblyFileVersionAttribute>(a => a.Version);
            }
        }

        // ----------------------------------------------------------------------
        // Version attributes
        // ----------------------------------------------------------------------
        public static Version Version
        {
            get
            {
                // ----------------------------------------------------------------------
                // Get the assembly, return empty if null
                // ----------------------------------------------------------------------
                var assembly = GetAssembly(StackTraceLevel);
                return assembly is null ? new Version() : assembly.GetName().Version;
            }
        }

        public string VersionFull
        {
            get
            {
                return Version.ToString();
            }
        }

        public string VersionMajor
        {
            get
            {
                return Version.Major.ToString();
            }
        }

        public string VersionMinor
        {
            get
            {
                return Version.Minor.ToString();
            }
        }

        public string VersionBuild
        {
            get
            {
                return Version.Build.ToString();
            }
        }

        public string VersionRevision
        {
            get
            {
                return Version.Revision.ToString();
            }
        }

        // ----------------------------------------------------------------------
        // Set how deep in the stack trace we're looking - allows for customized changes
        // ----------------------------------------------------------------------
        public static int StackTraceLevel
        {
            get
            {
                return m_StackTraceLevel;
            }
            set
            {
                m_StackTraceLevel = value;
            }
        }

        private static int m_StackTraceLevel;

        /// <summary>
        /// Gets the calling assembly attribute.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="value">The value.</param>
        /// <example>return GetCallingAssemblyAttribute&lt;AssemblyCompanyAttribute&gt;(a => a.Company);</example>
        /// <returns></returns>
        private string GetCallingAssemblyAttribute<T>(Func<T, string> value) where T : Attribute
        {
            // ----------------------------------------------------------------------
            // Get the assembly, return empty if null
            // ----------------------------------------------------------------------
            var assembly = GetAssembly(StackTraceLevel);
            if (assembly is null)
            {
                return string.Empty;
            }

            // ----------------------------------------------------------------------
            // Get the attribute value
            // ----------------------------------------------------------------------
            T attribute__1 = (T)Attribute.GetCustomAttribute(assembly, typeof(T));
            return value.Invoke(attribute__1);
        }

        /// <summary>
        /// Go through the stack and gets the assembly
        /// </summary>
        /// <param name="stackTraceLevel">The stack trace level.</param>
        /// <returns></returns>
        private static Assembly GetAssembly(int stackTraceLevel)
        {
            // ----------------------------------------------------------------------
            // Get the stack frame, returning null if none
            // ----------------------------------------------------------------------
            var stackTrace = new StackTrace();
            StackFrame[] stackFrames = stackTrace.GetFrames();
            if (stackFrames is null)
            {
                return null;
            }

            // ----------------------------------------------------------------------
            // Get the declaring type from the associated stack frame, returning null if nonw
            // ----------------------------------------------------------------------
            var declaringType = stackFrames[stackTraceLevel].GetMethod().DeclaringType;
            if (declaringType is null)
            {
                return null;
            }

            // ----------------------------------------------------------------------
            // Return the assembly
            // ----------------------------------------------------------------------
            var assembly = declaringType.Assembly;
            return assembly;
        }

        internal static string MethodName(string currentMethod)
        {
            var result = new StringBuilder();
            var stackTrace = new StackTrace();
            StackFrame[] stackFrames = stackTrace.GetFrames();
            int Index = stackFrames.Select((elem, indexInternal) => new { elem, indexInternal }).First(p => p.elem.GetMethod().Name.ToLower().Equals(currentMethod)).indexInternal + 1;
            if (stackFrames.IsNotEmpty() && stackFrames.Length >= Index)
            {
                var stackCall = stackFrames[Index];
                if (stackCall.GetMethod().IsNotEmpty() && stackCall.GetMethod().Name.IsNotEmpty())
                {
                    result.AppendLine(string.Format("<<I>>Method:'{0}'", stackCall.GetMethod().Name));
                    string nameSpace = string.Empty;
                    if (stackCall.GetMethod().ReflectedType.IsNotEmpty() && stackCall.GetMethod().ReflectedType.FullName.IsNotEmpty())
                    {
                        string className = stackCall.GetMethod().ReflectedType.FullName;
                        nameSpace = string.Format("{0}.{1}", className, stackCall.GetMethod().Name);
                    }
                    if (nameSpace.IsEmpty())
                    {
                        nameSpace = "Empty";
                    }
                    result.AppendLine(string.Format("<<I>>Namespace:'{0}'", nameSpace));
                }
                else
                {
                    result.AppendLine(string.Format("<<I>>Method:'{0}'", "Empty"));
                }
                result.AppendLine(string.Format("<<I>>Row/Column:'{0}'/'{1}'", stackCall.GetFileLineNumber(), stackCall.GetFileColumnNumber()));
            }
            return result.ToString();
        }

        public static string GetFrameProcess(int Index)
        {
            var result = new StringBuilder();
            var stackTrace = new StackTrace();
            StackFrame[] stackFrames = stackTrace.GetFrames();
            if (stackFrames.IsNotEmpty() && stackFrames.Length >= Index)
            {
                var stackCall = stackFrames[Index];
                if (stackCall.GetMethod().IsNotEmpty() && stackCall.GetMethod().Name.IsNotEmpty())
                {
                    result.AppendLine(string.Format("<<I>>Method:'{0}'", stackCall.GetMethod().Name));
                    string nameSpace = string.Empty;
                    if (stackCall.GetMethod().ReflectedType.IsNotEmpty() && stackCall.GetMethod().ReflectedType.FullName.IsNotEmpty())
                    {
                        string className = stackCall.GetMethod().ReflectedType.FullName;
                        nameSpace = string.Format("{0}.{1}", className, stackCall.GetMethod().Name);
                    }
                    if (nameSpace.IsEmpty())
                    {
                        nameSpace = "Empty";
                    }
                    result.AppendLine(string.Format("<<I>>Namespace:'{0}'", nameSpace));
                }
                else
                {
                    result.AppendLine(string.Format("<<I>>Method:'{0}'", "Empty"));
                }
                result.AppendLine(string.Format("<<I>>Row/Column:'{0}'/'{1}'", stackCall.GetFileLineNumber(), stackCall.GetFileColumnNumber()));
            }
            return result.ToString();
        }

        public static string GetFrameProcessFullName(int Index)
        {
            string result = string.Empty;
            var stackTrace = new StackTrace();
            StackFrame[] stackFrames = stackTrace.GetFrames();
            if (stackFrames.IsNotEmpty() && stackFrames.Length >= Index)
            {
                var stackCall = stackFrames[Index];
                if (stackCall.GetMethod().IsNotEmpty() && stackCall.GetMethod().Name.IsNotEmpty())
                {
                    result = stackCall.GetMethod().Name;
                    string nameSpace = string.Empty;
                    if (stackCall.GetMethod().ReflectedType.IsNotEmpty() && stackCall.GetMethod().ReflectedType.FullName.IsNotEmpty())
                    {
                        string className = stackCall.GetMethod().ReflectedType.FullName;
                        nameSpace = string.Format("{0}.{1}", className, stackCall.GetMethod().Name);
                    }
                    if (nameSpace.IsEmpty())
                    {
                        nameSpace = "Empty";
                    }
                    result = nameSpace;
                }
            }
            return result;
        }

        /// <summary>
        /// Obtiene el nombre del usuario de windows.
        /// </summary>
        /// <returns></returns>
        public static string GetUserNameWindows()
        {
            string result = string.Empty;
            if (!HostingEnvironment.IsHosted)
            {
                result = Environment.UserName;
            }
            return result;
        }

        /// <summary>
        /// Obtiene el nombre de la maquina.
        /// </summary>
        /// <returns></returns>
        public static string GetUserNameMachine()
        {
            string result = string.Empty;
            if (!HostingEnvironment.IsHosted)
            {
                result = Environment.MachineName;
            }
            return result;
        }

        /// <summary>
        /// Obtiene el nombre de la sistema operativo.
        /// </summary>
        /// <returns></returns>
        public static string GetOS()
        {
            string result = string.Empty;
            try
            {
                if (!HostingEnvironment.IsHosted)
                {
                    result = System.Runtime.InteropServices.RuntimeInformation.OSDescription;
                }
            }
            catch (Exception ex)
            {
                LogHandler.ErrorLog("InMotionGIT.Common.Helpers.AssemblyHandler", "GetOS", ex);
                result = "Unknown";
            }

            return result;
        }

        /// <summary>
        /// Obtiene el nombre de la sistema operativo.
        /// </summary>
        /// <returns></returns>
        public static string GetNameApplication()
        {
            string result = string.Empty;
            if (!HostingEnvironment.IsHosted)
            {
                try
                {
                    result = new AssemblyHandler(5).Title;
                    if (result.IsEmpty())
                    {
                        result = Assembly.GetCallingAssembly().GetName().Name;
                    }
                }
                catch (Exception ex)
                {
                    result = "Name not found";
                }
            }
            return result;
        }

        public static string GetVersionApplication()
        {
            string result = string.Empty;
            if (!HostingEnvironment.IsHosted)
            {
                try
                {
                    result = new AssemblyHandler(5).VersionFull;
                    if (result.IsEmpty())
                    {
                        result = Assembly.GetCallingAssembly().GetName().Version.ToString();
                    }
                }
                catch (Exception ex)
                {
                    result = "-1";
                }
            }
            return result;
        }

    }

}